### ISSUE TITLE
HTTP redirect to HTTPS for web UI

### ISSUE THREAD

**Issue by gtmills** [ibm-openbmc/dev#895]
On op920, op930 we had a redirect, via Nginx, from HTTP to HTTPS.
This was especially handy for the Web UI. 
We lost this function when moving to bmcweb. 
This story is to implement this functionality into bmcweb. 

From the mailing list: 
```
bmcweb can support non-TLS http via the BMCWEB_INSECURE* configuration
options, although as the name implies, they are insecure, and therefore
shouldn't be enabled for anything relating to production, but is useful
in a debug context, or to compare the performance effects of encryption
on the end result.

If what you've described is something you're wanting to enable, there's
a few options that have been discussed in the past.

1. An http redirect to https.  While not the most secure option, if
implemented with the correct security headers and in a way that can be
disabled by the end user for high security environments, it does improve
the user experience when using a browser.  I'd be in full support of
this being done, if done properly.
```

**Comment by rfrandse** [comment]
refresh

**Comment by rfrandse** [comment]
refresh

**Comment by rfrandse** [comment]
refresh

**Comment by joseph-reynolds** [comment]
Is this redirect only needed for specific URLs?  For example, for a BMC at IP address 192.168.10.20, you could type the following into the browser's address bar:
1. 192.168.10.20
2. http://192.168.10.20
3. https://192.168.10.20
4. https://192.168.10.20/index.html

Before the switch to BMCWeb, any of these would have worked.  Option 1 would have the "http://" added by the browser.  Options 1 and 2 would have the nginx web server redirect the browser to https.  Options 3 and 4 would retrieve the Web App.  ==> The question is: do you need any other URLs to redirect from http to https?

**Comment by joseph-reynolds** [comment]
I interpret Ed's suggestions as follows
1. Enable HTTP -> HTTPS redirect by default and have a new BMCWeb configuration option (https://github.com/openbmc/bmcweb/blob/master/CMakeLists.txt) to disable it.
2. Redirect correctly to avoid security issues, such as by following OWASP recommendations like https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html.  I believe BMCWeb already applies the required HTTP response headers.

**Comment by gtmills** [comment]
I think http://192.168.10.20,  http://yoursystem.com,  http://yoursystem.com/#/server-health/event-log should all redirect

**Comment by joseph-reynolds** [comment]
I pushed the design for this before reading your comment: https://gerrit.openbmc-project.xyz/c/openbmc/docs/+/24173
I can see the benefit of redirecting those URIs.  How will BMCWeb know which requests to redirect to HTTPS and which ones to trash?  Is there a pattern to the URIs?
Let's take this discussion to the review.

**Comment by joseph-reynolds** [comment]
I've added the HSTS spec to the design review.  The spec section 7.2 (redirect HTTP to HTTPS) is exactly what we need to implement in BMCWeb.  I didn't find any OWASP recommendations on this topic.

**Comment by joseph-reynolds** [comment]
What should http://mybmc.com/#/server-health/event-log redirect to?  Do you mean:
- https://mybmc.com/#/server-health/event-log or
- https://mybmc.com

**Comment by gtmills** [comment]
https://mybmc.com/#/server-health/event-log

**Comment by joseph-reynolds** [comment]
Two questions:
1. How would the user have gotten the URL http://mybmc.com/#/server-health/event-log in the first place?  Why would they ever enter it into their browser?
Let me back up up a step here ... My impression for having the http->https redirect is for users who type the hostname into the browser without the https:// prefix.  That's the only use case I think we need solve. 

2. If we redirect more than just / and /index.html, how will BMCWeb know which URIs to redirect?  If it redirects all of them, then it will redirect http://HOST/redfish/v1/SessionManager/Session/... to https://HOST/redfish/v1/SessionManager/Session/...  which represents a security risk (someone POSTing their credentials via insecure HTTP).

**Comment by gtmills** [comment]
@joseph-reynolds I don't think it is critical we redirect  http://mybmc.com/#/server-health/event-log  to https://mybmc.com/#/server-health/event-log  instead of https://mybmc.com. This is how the nginx server redirect worked though.

**Comment by gtmills** [comment]
I do agree that this is not common the user would go to http://mybmc.com/#/server-health/event-log  but could happen if the user types in the url, copies a link, etc.

**Comment by joseph-reynolds** [comment]
The latest design is in https://gerrit.openbmc-project.xyz/c/openbmc/docs/+/24173/2/designs/http-redirect.md
Please note the review comments which go with that design have critical information but have not yet been incorporated into the design doc.
The work at this point is to update that design to incorporate the comments, notify the community that we intend to implement this design, and then do the work.  It should all be in bmcweb.

**Comment by lxwinspur** [comment]
@gtmills 
I took a cursory look at the context and design doc and did some testing.
I have a couple of questions:
1. The most important is that the current communication logic can only listen to one communication protocol (http or https). If you listen to both protocols at the same time, you need to change a lot of code and communication logic, which will cause some unnecessary troubles.
2. Could be listen to ports 443 and 80 at the same time in the bmcweb.socket file, or add another file?
3. Why not continue to use nginx? What's the reason?

**Comment by gtmills** [comment]
2) Could work. I would run your proposal by the BMC maintainers via the emailing list before implementing it. 
3) To save flash space, we removed nginx. Nginx seems like overkill just for a http to https redirect but I personally would not be opposed. @geissonator @joseph-reynolds

**Comment by mine260309** [comment]
From my point of view, I prefer to use nginx for such case:
1. nginx is the perfect tool to get this feature for free, only a proper config file is needed.
2. nginx could be configured to remove unnecessary features at compile time, not sure the final binary size though.
3. If we are going to implement this feature in bmcweb, it costs extra effort and it's likely the implementation is no better than nginx.

Anyway, such discussion is better to be discussed in the mailing list, is it?

**Comment by amboar** [comment]
Rather than pull in nginx, why not just use `nc` and a shell script. Depending on how it's done `nc` may even be unnecessary. We just need to issue a 308 permanent redirect response.

**Comment by lxwinspur** [comment]
review:
https://gerrit.openbmc-project.xyz/c/openbmc/meta-openpower/+/28015
https://gerrit.openbmc-project.xyz/c/openbmc/meta-openpower/+/28099

**Comment by lxwinspur** [comment]
The new commit address is https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-misc/+/28353, and I will follow up on this commit.

**Comment by mdmillerii** [comment]
The commit is ready for review at https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-misc/+/28353 along with meson and systemd units in https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-misc/+/31166 and additional unit tests in https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-misc/+/31167 .  While I have added some self review comments they are more future enhancements and I am waiting for comments from others.

**Comment by lxwinspur** [comment]
Hi, @gtmills @mzipse 
This task has been finished by @mdmillerii(the above three commits has been merged), so could you close it if you have not any question.
thanks :)

**Comment by mdmillerii** [comment]
There is ongoing discussion on the compeeting proposal for bmcweb see https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/35265

**Comment by anoo1** [comment]
> There is ongoing discussion on the compeeting proposal for bmcweb see https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/35265

It appears this commit is close to being merged, so we can enable this bmcweb option instead of pulling in the phosphor-misc redirect package.

**Comment by gtmills** [comment]
> gtmills modified the milestones: A.1.999, A.1.213.9 now 

I think we want this for GA. Moved accordingly.

**Comment by rfrandse** [comment]
tracking in jira
